[TNF] Update specifics of etcd setup, TNF job, and TNF controller. (#12)
mshitrit/enhancements: enhancements/two-node-fencing/tnf.md

| ----------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- |
| [Feature Gates](#feature-gate-changes)                            | Add a new `DualReplicaTopology` feature which can be enabled via the `CustomNoUpgrade` feature set                              |
| [OpenShift API](#openshift-api-changes)                           | Add `DualReplica` as a new value for `ControlPlaneTopology`                                                                     |

> | [ETCD Operator](#etcd-operator-changes)                           | Add a mode to stop managing the etcd container, a new scaling strategy, and a operator + controller for initializing pacemaker  |

Reviewer: since we very probably switch from running the pacemaker code in a Deployment, to run it in a Job, I think that doesn't qualify as "operator" anymore. What about using something more generic, e.g. "... and a workload with a controller for initializing pacemaker" 🤔

An important facility of the installation flow is the transition from a CEO deployed etcd to one controlled by RHEL-HA. The basic transition works as follows:
1. [MCO extensions](https://docs.openshift.com/container-platform/4.17/machine_configuration/machine-configs-configure.html#rhcos-add-extensions_machine-configs-configure) are used to ensure that the
   pacemaker and corosync RPMs are installed. The installer also creates MachineConfig manifests to pre-configure resource agents.

> 2. Upon detection that the cluster infrastructure is using the DualReplica controlPlaneTopology in the infrastructure config, a TNF controller running in CEO will run commands on one of the cluster

Reviewer: "... a TNF controller **started by** CEO.. "

   nodes to initialize pacemaker. The outcome of this is that the podman-etcd resource agent will be started on both nodes.
3. The aforementioned TNF controller will signal CEO to relinquish control of etcd by setting CEO's `managedEtcdKind` to `External`. When this happens, CEO immediately removes the etcd container from
   the static pod configs. The resource agents for etcd are running from step 2, and they are configured to wait for etcd containers to be gone so they can restart them using Podman.
4. The installation proceeds as normal once the containers start. If for some reason, the etcd containers cannot be started, then the installation will fail. The installer will pull logs from the
control-plane nodes to provide context for this failure.

###### Managing Pacemaker and Resource/Fence agent Configuration
Pacemaker configurations (as well as its resource and fence agent configurations) do not need to be stored as files, as they are dynamically created using pcs commands run by new controller called the

> TNF contoller.

Reviewer 2: typo `s/contoller/controller`

See the API Extensions section below for sample install-configs.

For a two-node cluster to be successful, we need to ensure the following:

> 1. The BMC secrets for RHEL-HA will be in a new section of the install-config.yaml. These are consumed by the default flow of creating manifests and rendering new secrets in the openshift-etcd

Reviewer 2: nitpick, but this seems a bit hard to follow, what about?

```suggestion
1. The BMC secrets for RHEL-HA, located in a new section of the install-config.yaml, are consumed by the default flow of creating manifests and rendering new secrets in the openshift-etcd
```

#### etcd Operator Changes

Initially, the creation of an etcd cluster will be driven in the same way as other platforms. Once the cluster has two members, the etcd daemon will be removed from the static pod definition and

> recreated as a resource controlled by RHEL- a resource agent called podman-etcd. At this point, the cluster-etcd-operator (CEO) will be made aware of this change so that some membership management

Reviewer 2: is this wanted, or the "RHEL-" part was also intended to be removed?

Author: I think something happened by accident. RHEL-HA was meant to stay :)

functionality that is now handled by RHEL-HA can be disabled. This will be achieved by the TNF controller running under CEO by setting a field in the CEO's `ConfigMap` - which can only succeed if the
control-plane is healthy.

To enable this flow, we propose the addition of a `managedEtcdKind` field which defaults to `Cluster` but will be set to `External` during installation, and will only be respected if the
`Infrastructure` CR's `TopologyMode` is `DualReplicaTopologyMode`. This will allow the use of a credential scoped to `ConfigMap`s in the `openshift-etcd-operator` namespace, to make the change.

While set to `External`, CEO will still need to render the configuration for the etcd container to be consumed by pacemaker. This ensures that the etcd instance managed by pacemaker's podman-etcd
resource agent can be updated accordingly in the case of an upgrade event or whenever certificates are rotated.

##### The TNF Operator and TNF Controller
Pacemaker initialization is initiated by a new TNF operator, which is created by CEO when it detects that the cluster controlPlane topology is set to `DualReplica`. This operator exists for two main

> reasons - a separation of concerns focused around TNF initialization (which is specific to the Two-Node with Fencing topology), and the management of a new namespace with privileges capable of running

Reviewer: > management of a new namespace

in review we were asked to use openshift-etcd instead of a new one, which works because it has the needed `pod-security.kubernetes.io/enforce: privileged` label

This controller is called the TNF controller, and its responsibility is to invoke these `pcs` commands and ensure CEO ends up in a degraded state if they are unsuccessful. You can think of this
controller as the component responsible for ensuring that etcd is healthy and set up to properly handle split-brain and node-failure events in TNF clusters.

The TNF controller requires privileged mode because it will call the `pcs` utility to configure the pacemaker cluster (which needs elevated permissions). CEO has the `cluster-admin` RBAC role, but it
doesn't run in privileged mode or allow privileged workloads (i.e. the pod security admission is set to `restricted` via a namespace annotation). This necessitates the introduction of a new operator
to run this controller - one that runs in a namespace with the `privileged` annotation, running in privileged mode, and has hostPID = true.

A full discussion of the the the CEO changes, as well as alternatives to this this approach, is available in our [TNF in CEO design
document](https://docs.google.com/document/d/1ZqRYPM2OZ6kdD_tua4lB-gf3iSL-cuAoicvqIbM10MQ/edit?usp=sharing).

> In the future, it may be possible to lower the privilege level of the TNF controller, once the the RHEL HA team can rearchitect `pcs` to a full client-server architecture which would allow the `pcs`

Reviewer 2: typos
* `s/the the/the/`
* `s/rearchitect/rearchitects`
